cli_tools/gce_export/main.go

```diff
@@ -69,99 +91,247 @@ func splitLicenses(input string) []string {
 	return ls
 }
 
-func main() {
-	flag.Parse()
+type buffer struct {
+	// These fields are read only.
+	cSize    int64
+	prefix   string
+	ctx      context.Context
+	id       string
+	bkt, obj string
 
-	if *localPath != "" && *gcsPath != "" {
-		log.Fatal("-local_path and -gcs_path can't be both specified")
-	}
+	upload    chan string
+	tmpObjs   []string
+	tmpObjsMx sync.Mutex
 
-	if *localPath == "" && *gcsPath == "" {
-		log.Fatal("-local_path or -gcs_path must be provided")
-	}
+	sync.Mutex
+	sync.WaitGroup
+	bytes int64
+	part  int
+	file  *os.File
+}
 
-	if *disk == "" {
-		log.Fatal("The flag -disk must be provided")
+func (b *buffer) addObj(obj string) {
+	b.tmpObjsMx.Lock()
+	b.tmpObjs = append(b.tmpObjs, obj)
+	b.tmpObjsMx.Unlock()
+}
+
+func (b *buffer) uploadWorker() {
+	defer b.Done()
+	for in := range b.upload {
+		for i := 1; ; i++ {
+			err := func() error {
+				client, err := gcsClient(b.ctx)
+				if err != nil {
+					return err
+				}
+				defer client.Close()
+
+				file, err := os.Open(in)
+				if err != nil {
+					return err
+				}
+				defer file.Close()
+				defer os.Remove(in)
+
+				tmpObj := path.Join(b.obj, strings.TrimPrefix(in, b.prefix))
+				b.addObj(tmpObj)
+				dst := client.Bucket(b.bkt).Object(tmpObj).NewWriter(b.ctx)
```

Comment:
per my understanding, if an error is thrown on the half way, there will be an incomplete file left on GCS. I am not sure whether GCS allow us to ignore that file and simply redo uploading. Even if it is allowed, it would be better to resume, while we are able to tracked the progress. If that is hard to implement, we can leave it as a improvement in the future.

Reply:
The file isn't written before Close is called, we attempt to reupload this bit up to retry times, if the entire process is canceled then yes you have temporary files left around. We could maybe add logic to try to find old temporary bits and check if they match so we don't re upload but that won't really help our process as this is all done in a workflow that we can't really replicate on additional runs.

Reply:
If "The file isn't written before Close is called" is correct, then I totally agree all the statements. I am still not 100% sure about that because I failed to find a official description about this. I will do more test after your PR got release, and improve it if needed.